x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/PCG.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs;
+
+import com.carrotsearch.hppc.BitMixer;
+
+/**
+ * Permuted congruential generator style random number generator
+ */
+public final class PCG {
```

Comment:
With random number generators there are some pretty nice benchmarking tools available. PCG tends to perform extremely well on all tests of statistical properties of rng: BigCrush, PractRand, etc. (DieHarder, mentioned in `SpittableRandom` documentation, is generally considered a little weak.) It is also extremely fast and small state and is used by some high quality project, for example it's the default RNG for numpy IIRC. There is a lot more information [here](https://www.pcg-random.org/pdf/hmc-cs-2014-0905.pdf).

That said this was mainly used by the various sampling components of this PR because they were originally prototyped C++ and it inherited the rng we were using there. It could be that `SplittableRandom` is fine for our needs. Essentially, we'd like high quality random numbers as fast as possible and the ability to generate streams which are statistically independent. Maybe this is something we could benchmark.

Reply:
I can try some benchmarks locally, but what I found is that all other random number generators ended up being slower.

Reply:
Also, it is useful to have one seed and have the ability to create non-correlative streams from the one seed. I know this is possible with other random number generators, but maybe not as easily.

Reply:
Ah, looking at `SplittableRandom.split()` it effectively does this. It mentions that the random streams share the same statistical distribution (with high probability), but the two streams will not be correlated.

I will try it out, looks promising (didn't really know about it).